File: android/app/src/main/java/com/canvasleaf/app/engine/WikiLinkExtractor.kt
package com.canvasleaf.app.engine

import com.canvasleaf.app.model.CanvasEdge
import com.canvasleaf.app.model.CanvasNode

/**
 * WikiLink Extractor Engine (SDD Section 4.2).
 * Parses `[[NoteName]]` or `[[relation]] NoteName` or `[[Target|Alias]]`
 * from markdown content to build an adjacency graph of nodes and edges automatically.
 */
object WikiLinkExtractor {
    private val WIKILINK_REGEX = Regex("""\[\[(.*?)\]\]""")
    private val LABELED_LINK_REGEX = Regex("""\[\[([a-zA-Z0-9_\s-]+)\]\]\s*(?:(?:->|-->|──>|─►|—►|-►)\s*|(?:\s*-\s*))?\[\[(.*?)\]\]""")
    private val PREFIX_LABEL_REGEX = Regex("""\[\[([a-zA-Z0-9_\s-]+)\]\]\s+([a-zA-Z0-9_\-\.]+)""")

    data class ExtractedLink(
        val targetTitle: String,
        val label: String? = null
    )

    fun extractLinks(markdown: String): List<ExtractedLink> {
        val links = mutableListOf<ExtractedLink>()
        val seen = mutableSetOf<String>()

        // 1. Check labeled pattern: [[uses]] [[SessionDB]]
        LABELED_LINK_REGEX.findAll(markdown).forEach { match ->
            val label = match.groupValues[1].trim()
            val target = match.groupValues[2].trim().removeSuffix(".md")
            val key = "$target:$label"
            if (seen.add(key)) {
                links.add(ExtractedLink(targetTitle = target, label = label))
            }
        }

        // 2. Check prefix pattern: [[uses]] SessionDB.md
        PREFIX_LABEL_REGEX.findAll(markdown).forEach { match ->
            val label = match.groupValues[1].trim()
            val target = match.groupValues[2].trim().removeSuffix(".md")
            val key = "$target:$label"
            if (seen.add(key)) {
                links.add(ExtractedLink(targetTitle = target, label = label))
            }
        }

        // 3. Check regular [[Target]] or [[Target|Alias]]
        WIKILINK_REGEX.findAll(markdown).forEach { match ->
            val inner = match.groupValues[1].trim()
            val parts = inner.split("|")
            val target = parts[0].trim().removeSuffix(".md")
            val alias = if (parts.size > 1) parts[1].trim() else null

            val key = "$target:${alias ?: ""}"
            if (seen.add(key)) {
                links.add(ExtractedLink(targetTitle = target, label = alias))
            }
        }

        return links
    }

    /**
     * Constructs a directed graph of edges across all loaded nodes
     */
    fun buildAdjacencyGraph(nodes: List<CanvasNode>): List<CanvasEdge> {
        val edges = mutableListOf<CanvasEdge>()
        val nodeMap = mutableMapOf<String, CanvasNode>()

        nodes.forEach { node ->
            nodeMap[node.id.lowercase()] = node
            nodeMap[node.title.lowercase()] = node
            val baseName = node.filePath.substringAfterLast("/").substringAfterLast("\\").removeSuffix(".md").lowercase()
            nodeMap[baseName] = node
        }

        val createdPairs = mutableSetOf<String>()

        nodes.forEach { sourceNode ->
            val extracted = extractLinks(sourceNode.contentSnippet)
            extracted.forEach { link ->
                val targetNode = nodeMap[link.targetTitle.lowercase()]
                if (targetNode != null && targetNode.id != sourceNode.id) {
                    val pairKey = "${sourceNode.id}->${targetNode.id}"
                    if (createdPairs.add(pairKey)) {
                        edges.add(
                            CanvasEdge(
                                edgeId = "edge-${sourceNode.id}-${targetNode.id}",
                                fromNodeId = sourceNode.id,
                                toNodeId = targetNode.id,
                                label = link.label
                            )
                        )
                    }
                }
            }
        }

        return edges
    }
}

File: web/src/engines/WikiLinkExtractor.ts
import { CanvasEdge, CanvasNode } from '../types/canvas';

export interface ExtractedLink {
  targetTitle: string;
  label?: string;
  isBiDirectional?: boolean;
}

export class WikiLinkExtractor {
  /**
   * Extract all outbound links from markdown text
   */
  public static extractLinks(markdown: string): ExtractedLink[] {
    const links: ExtractedLink[] = [];
    const seen = new Set<string>();

    // 1. Check for labeled pattern like: [[uses]] [[SessionDB]] or [[uses]] SessionDB.md
    const labeledMatches = markdown.matchAll(/\[\[([a-zA-Z0-9_\s-]+)\]\]\s*(?:(?:->|-->|──>|─►|—►|-►)\s*|(?:\s*-\s*))?\[\[([a-zA-Z0-9_\-\.\s]+)\]\]/g);
    for (const m of labeledMatches) {
      const label = m[1].trim();
      const target = m[2].trim().replace(/\.md$/i, '');
      const key = `${target}:${label}`;
      if (!seen.has(key)) {
        seen.add(key);
        links.push({ targetTitle: target, label });
      }
    }

    // Pattern: [[uses]] TargetName.md (without second brackets)
    const relPrefixMatches = markdown.matchAll(/\[\[([a-zA-Z0-9_\s-]+)\]\]\s+([a-zA-Z0-9_\-]+(?:\.md)?)/g);
    for (const m of relPrefixMatches) {
      const label = m[1].trim();
      const target = m[2].trim().replace(/\.md$/i, '');
      const key = `${target}:${label}`;
      if (!seen.has(key)) {
        seen.add(key);
        links.push({ targetTitle: target, label });
      }
    }

    // 2. Standard [[Target]] or [[Target|Alias]] or [[Target]] - label
    const lines = markdown.split('\n');
    for (const line of lines) {
      const lineMatches = line.matchAll(/\[\[(.*?)\]\]/g);
      for (const m of lineMatches) {
        let inside = m[1].trim();
        let label: string | undefined = undefined;

        // Check if it has an alias [[Target|Alias]]
        if (inside.includes('|')) {
          const parts = inside.split('|');
          inside = parts[0].trim();
          label = parts[1].trim();
        }

        const cleanTarget = inside.replace(/\.md$/i, '');
        
        // Check if there's a relation label in the line: e.g. [[Target]] - sync or [[Target]] - [[sync]]
        const afterMatch = line.substring((m.index || 0) + m[0].length).trim();
        const labelMatch = afterMatch.match(/^(?:-\s*|:\s*|—\s*|──►\s*|─►\s*)(?:\[\[([a-zA-Z0-9_\s-]+)\]\]|([a-zA-Z0-9_\s-]+))/);
        if (labelMatch) {
          label = (labelMatch[1] || labelMatch[2] || '').trim();
        }

        const key = `${cleanTarget}:${label || ''}`;
        if (!seen.has(key)) {
          seen.add(key);
          links.push({ targetTitle: cleanTarget, label });
        }
      }
    }

    return links;
  }

  /**
   * Build graph edges automatically across all canvas nodes
   */
  public static buildEdgesFromNodes(nodes: CanvasNode[]): CanvasEdge[] {
    const edges: CanvasEdge[] = [];
    const nodeMap = new Map<string, CanvasNode>();

    // Normalize map lookup by id, title, and base filename
    for (const node of nodes) {
      nodeMap.set(node.id.toLowerCase(), node);
      nodeMap.set(node.title.toLowerCase(), node);
      const baseName = node.filePath.split(/[\\/]/).pop()?.replace(/\.md$/i, '').toLowerCase();
      if (baseName) {
        nodeMap.set(baseName, node);
      }
    }

    const createdEdgeKeys = new Set<string>();

    for (const sourceNode of nodes) {
      const links = this.extractLinks(sourceNode.contentSnippet);
      for (const link of links) {
        const targetLookupKey = link.targetTitle.toLowerCase();
        const targetNode = nodeMap.get(targetLookupKey);

        if (targetNode && targetNode.id !== sourceNode.id) {
          const edgeKey = `${sourceNode.id}->${targetNode.id}`;
          if (!createdEdgeKeys.has(edgeKey)) {
            createdEdgeKeys.add(edgeKey);
            edges.push({
              edgeId: `edge-${sourceNode.id}-${targetNode.id}-${edges.length}`,
              fromNodeId: sourceNode.id,
              toNodeId: targetNode.id,
              label: link.label || null,
              isBiDirectional: false,
            });
          }
        }
      }
    }

    return edges;
  }
}
